blog/_posts/2019-09-06-profilers.md

```diff
@@ -0,0 +1,405 @@
+
+```
+$ perf report --stdio --call-graph=none
+# Total Lost Samples: 0
+#
+# Samples: 4K of event 'cycles:uppp'
+# Event count (approx.): 3344003009
+#
+# Children      Self  Command  Shared Object     Symbol
+# ........  ........  .......  ................  ..................................
+#
+    64.22%    64.22%  a.out    a.out             [.] fib
+    43.30%    35.77%  a.out    a.out             [.] fib_r
+     8.75%     0.00%  a.out    [unknown]         [.] 0x00007ffc2454634f
+     7.52%     0.00%  a.out    a.out             [.] _start
+     7.52%     0.00%  a.out    a.out             [.] main
+```
+
+I tried various options including `-gdwarf-2`, `-gdwarf-4`, `-fno-omit-frame-pointer`, and switching to `clang`. None of those helped much.
+With FPO disabled, the clang results were much better than with gcc (both improved), but clang's output was also a smidgen slower running,
+so that might be a contributing factor (more overhead).
+
+But combining the compiler option `-fno-omit-frame-pointer` with the perf option `--call-graph fp` worked.
+
+```
+$ perf report --stdio --call-graph=none
+# Total Lost Samples: 0
+#
+# Samples: 4K of event 'cycles:uppp'
+# Event count (approx.): 2752040208
+#
+# Children      Self  Command  Shared Object     Symbol
+# ........  ........  .......  ................  ..........................
+#
+    99.99%    65.02%  a.out    a.out             [.] fib
+    99.99%    34.97%  a.out    a.out             [.] fib_r
+    99.99%     0.00%  a.out    a.out             [.] main
+    99.99%     0.00%  a.out    libc-2.27.so      [.] __libc_start_main
+    99.99%     0.00%  a.out    [unknown]         [.] 0x077e258d4c544155
+```
+
+Here we see it's got the right children (total) and self weights!
+
+There's also a bunch of other options for `--call-graph`, but all of them spewed vast amounts of (mis)information, so I've opted not to try to fit that in here.
+
+## Profiling JuliaLang code
+
+Now let's see how JuliaLang (v1.3) does:
+
+```julia
+julia> @noinline function fib(n)
+         return n > 1 ? fib_r(n - 1) + fib_r(n - 2) : 1
+       end
+
+julia> @noinline fib_r(n) = fib(n)
+
+julia> @time fib(40)
+  0.738735 seconds (3.16 k allocations: 176.626 KiB)
+165580141
+
+julia> using Profile
+
+julia> @profile fib(40)
+165580141
+
+julia> Profile.print(format=:flat, sortedby=:count)
+ Count File      Line Function
+    12 int.jl      52 -
+    14 int.jl      53 +
+   212 boot.jl    330 eval
+  5717 REPL[2]      1 fib_r
+  6028 REPL[1]      2 fib
+
+julia> count(==(0), Profile.fetch())
+585
+```
+
+So that's pretty bad—we have about 212 samples, and about 2000% of them were in `fib`. Clearly not true. What happened here? We computed that `fib` (and `fib_r`) each appeared an average of 10-20 times in every backtrace, so we over-represent them in our flat-format output. OK, so first thing to do is to fix that: each frame should only be counted once towards each output line.
```

Comment:
Might want to explain where the 212 comes from---why is `eval` a reliable indicator of the actual number of samples?